File: services/agent/app/models/providers.py
from langchain_core.language_models.chat_models import BaseChatModel
from langchain_openai import ChatOpenAI
from langchain_anthropic import ChatAnthropic
from langchain_google_genai import ChatGoogleGenerativeAI
import structlog

logger = structlog.get_logger(__name__)

def resolve_model(model_id: str, api_keys: dict) -> BaseChatModel:
    if not model_id:
        model_id = "openai/gpt-4o"

    parts = model_id.split("/")
    if len(parts) == 2:
        provider = parts[0]
        model_name = parts[1]
    else:
        provider = "openai"
        model_name = model_id

    logger.info("resolving_model", provider=provider, model_name=model_name)

    if provider == "openai":
        api_key = api_keys.get("openai")
        if not api_key:
            raise ValueError("OpenAI API key is missing. Please configure it in settings.")
        return ChatOpenAI(model=model_name, api_key=api_key)

    elif provider == "anthropic":
        api_key = api_keys.get("anthropic")
        if not api_key:
            raise ValueError("Anthropic API key is missing. Please configure it in settings.")
        return ChatAnthropic(model=model_name, api_key=api_key)

    elif provider == "google":
        api_key = api_keys.get("google")
        if not api_key:
            raise ValueError("Google API key is missing. Please configure it in settings.")
        return ChatGoogleGenerativeAI(model=model_name, google_api_key=api_key)

    elif provider == "groq":
        api_key = api_keys.get("groq")
        if not api_key:
            raise ValueError("Groq API key is missing. Please configure it in settings.")
        return ChatOpenAI(model=model_name, api_key=api_key, base_url="https://api.groq.com/openai/v1")

    elif provider == "mistral":
        api_key = api_keys.get("mistral")
        if not api_key:
            raise ValueError("Mistral API key is missing. Please configure it in settings.")
        return ChatOpenAI(model=model_name, api_key=api_key, base_url="https://api.mistral.ai/v1")

    elif provider == "ollama":
        base_url = api_keys.get("ollama", "http://localhost:11434/v1")
        return ChatOpenAI(model=model_name, api_key="ollama", base_url=base_url)

    else:
        raise ValueError(f"Unknown provider: {provider}")

File: src/lib/agent/providers.ts
import { ChatAnthropic } from "@langchain/anthropic";
import { ChatOpenAI } from "@langchain/openai";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";

export function resolveModel(model: string, apiKeys: Record<string, string>): BaseChatModel {
  const [provider, ...rest] = model.split("/");
  const modelName = rest.join("/");

  switch (provider) {
    case "anthropic": {
      const apiKey = apiKeys["anthropic"] || process.env.ANTHROPIC_API_KEY;
      if (!apiKey) throw new Error("No Anthropic API key configured");
      return new ChatAnthropic({
        model: modelName || "claude-sonnet-4-20250514",
        apiKey,
      });
    }
    case "openai": {
      const apiKey = apiKeys["openai"] || process.env.OPENAI_API_KEY;
      if (!apiKey) throw new Error("No OpenAI API key configured");
      return new ChatOpenAI({
        model: modelName || "gpt-4o",
        apiKey,
      });
    }
    case "groq": {
      const apiKey = apiKeys["groq"] || process.env.GROQ_API_KEY;
      if (!apiKey) throw new Error("No Groq API key configured");
      return new ChatOpenAI({
        model: modelName || "llama-3.3-70b-versatile",
        apiKey,
        configuration: { baseURL: "https://api.groq.com/openai/v1" },
      });
    }
    case "google": {
      const apiKey = apiKeys["google"] || process.env.GOOGLE_GENERATIVE_AI_API_KEY;
      if (!apiKey) throw new Error("No Google API key configured");
      return new ChatOpenAI({
        model: modelName || "gemini-2.0-flash",
        apiKey,
        configuration: { baseURL: "https://generativelanguage.googleapis.com/v1beta/openai/" },
      });
    }
    case "mistral": {
      const apiKey = apiKeys["mistral"] || process.env.MISTRAL_API_KEY;
      if (!apiKey) throw new Error("No Mistral API key configured");
      return new ChatOpenAI({
        model: modelName || "mistral-large-latest",
        apiKey,
        configuration: { baseURL: "https://api.mistral.ai/v1" },
      });
    }
    case "openrouter": {
      const apiKey = apiKeys["openrouter"] || process.env.OPENROUTER_API_KEY;
      if (!apiKey) throw new Error("No OpenRouter API key configured");
      return new ChatOpenAI({
        model: modelName || "openai/gpt-4o",
        apiKey,
        configuration: { baseURL: "https://openrouter.ai/api/v1" },
      });
    }
    default: {
      const apiKey = apiKeys[provider] || process.env.OPENAI_API_KEY;
      return new ChatOpenAI({ model: modelName || model, apiKey });
    }
  }
}
